platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java

```diff
@@ -65,32 +98,85 @@ public void close() throws Exception {
 
   @Override
   public void store(String pid, Dictionary properties) throws IOException {
-    ConfigurationContextImpl context = createContext(pid, properties);
-    if (context.shouldBeVisibleToPlugins()) {
-      List<ConfigurationPersistencePlugin> plugins =
-          asPlugins(configPersistenceTracker.getServices());
-      for (ConfigurationPersistencePlugin plugin : plugins) {
-        plugin.handleStore(context);
+    // Read lock to ensure plugin initialization is not currently occuring
+    long stamp = serviceStartingLock.readLock();
+    try {
+      ConfigurationContextImpl context = createContext(pid, properties);
+      if (context.shouldBeVisibleToPlugins()) {
+        List<ConfigurationPersistencePlugin> plugins =
+            asPlugins(configPersistenceTracker.getServices());
+        for (ConfigurationPersistencePlugin plugin : plugins) {
+          plugin.handleStore(context);
+        }
       }
+      super.store(pid, properties);
+    } finally {
+      serviceStartingLock.unlockRead(stamp);
     }
-    super.store(pid, properties);
   }
 
   @Override
   public void delete(String pid) throws IOException {
-    List<ConfigurationPersistencePlugin> plugins =
-        asPlugins(configPersistenceTracker.getServices());
-    for (ConfigurationPersistencePlugin plugin : plugins) {
-      plugin.handleDelete(pid);
+    // Read lock to ensure plugin initialization is not currently occuring
+    long stamp = serviceStartingLock.readLock();
+    try {
+      List<ConfigurationPersistencePlugin> plugins =
+          asPlugins(configPersistenceTracker.getServices());
+      for (ConfigurationPersistencePlugin plugin : plugins) {
+        plugin.handleDelete(pid);
+      }
+      super.delete(pid);
+    } finally {
+      serviceStartingLock.unlockRead(stamp);
     }
-    super.delete(pid);
   }
 
-  /* Factory method visible for testing purposes only */
+  /* Factory method visible for testing purposes - to inject a mock for the context */
   ConfigurationContextImpl createContext(String pid, Dictionary props) {
     return new ConfigurationContextImpl(pid, props);
   }
 
+  /* Factory method visible for testing purposes - to inject a mock for the plugin */
+  ConfigurationPersistencePlugin retrieveServiceObject(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    return FrameworkUtil.getBundle(DelegatingPersistenceManager.class)
+        .getBundleContext()
+        .getService(serviceReference);
+  }
+
+  private ConfigurationPersistencePlugin handleServiceAdded(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    ConfigurationPersistencePlugin plugin = retrieveServiceObject(serviceReference);
+    // Write lock because no configs should be getting saved or deleted while we initialize
+    long stamp = serviceStartingLock.writeLock();
+    try {
+      Enumeration<Dictionary<String, Object>> dictionaries =
+          getInnerPersistenceManager().getDictionaries();
```

Comment:
It should be noted that this is a very expensive operation. I think it's fine realistically because there will never be several `ConfigurationPersistencePlugin`s. But it may be worth making a tech debt ticket to address it.

Reply:
Not sure it should be addressed any other ways. The idea is to avoid going back to cfg admin which can lead to deadlock issues or lock complications. Since the persistent manager already manages all updates, relying on its own cache (i.e. the data/cache directory) is clean. To optimize this, we would need to create yet another cache and in this case, as you mentioned, there will never be many of these and it will typically happen only at config time. As such, this implementation is more than sufficient and I don't think we should create a tech debt ticket.

Reply:
I agree that I cannot think of a better solution. I just wanted to mention it. It may be worth noting somewhere in the javadoc at least then. Maybe on the `DelegatingPersistenceManager` as well as the `ConfigurationPersistencePlugin` interface?

Reply:
Just to warn future developers in this code that adding a `ConfigurationPersistencePlugin` should not be done lightly.

Reply:
👍  on adding a comment in the code

Reply:
Even if we had 12 config persistence plugins, that would translate to (at worst) an extra 12 - 24 seconds of container boot time (assuming it takes 1 -2 seconds to read the entire cache from disk). This only impacts plugin boot time, not plugin performance time in the heat of production. I'll provide background context, but I think this is trivial. 

Comments added inline and in javadoc.

Reply:
When I open the discussion with Felix config admin maintainers about possibly merging our solutions / framework improvements, I'll be sure to mention interoperability with a proper caching solution, such as Apache Ignite.

Reply:
@Lambeaux If you had 12 persistence plugin, then I would suggest not having a serial lock. Since the lock is designed to prevent updates while you take a snapshot, you could easily have the second persistence plugin starts its initialization at the same time. So detecting the lock is already acquired and actually wait for the same caching result to initialize as opposed to reloading it. Thus you would still have a 1-2 seconds delay unless it happens after the first one is done initializing. There is ways to even optimize this by temporarily caching stuff instead of doing it throughout runtime. There is a point where new persistence plugins should no longer be initialized and caching would not be required. In any case, this is moot right now so just a comment that if it ever happens, we should do something about it or better, a warning generated if a second plugin is registered might be something nice to get in the logs.